File: agid/xivo_agid/modules/schedule.py
__version__ = "$Revision$ $Date$"
__license__ = """
    Copyright (C) 2006, 2007, 2008  Proformatique

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA..
"""

from xivo_agid import agid

def forgetimefield(start, end):
	if start == '*':
		return '*'
	else:
		if end in (None, ''):
			return '%s' % start
		else:
			return '%s-%s' % (start, end)

def forgetime(res):
	return ','.join((
		forgetimefield(res['timebeg'], res['timeend']),
		forgetimefield(res['daynamebeg'], res['daynameend']),
		forgetimefield(res['daynumbeg'], res['daynumend']),
		forgetimefield(res['monthbeg'], res['monthend']),
	))

def schedule(handler, agi, cursor, args):
	id = args[0]

	cursor.query("SELECT ${columns} FROM schedule "
		     "WHERE id = %s "
		     "AND linked = 1 "
		     "AND commented = 0",
		     ('timebeg', 'timeend', 'daynamebeg', 'daynameend', 'daynumbeg', 'daynumend', 'monthbeg', 'monthend', 'typetrue', 'typevaltrue', 'applicationvaltrue', 'typefalse', 'typevalfalse', 'applicationvalfalse'),
		     (id,))
	res = cursor.fetchone()

	if not res:
		agi.dp_break("Invalid schedule ID '%s'" % id)

	agi.set_variable("XIVO_SCHEDULE_TIMERANGE", forgetime(res))
	handler.set_fwd_vars(res['typetrue'], res['typevaltrue'], res['applicationvaltrue'], 'XIVO_SCHEDULE_TYPETRUE', 'XIVO_SCHEDULE_TYPEVAL1TRUE', 'XIVO_SCHEDULE_TYPEVAL2TRUE')
	handler.set_fwd_vars(res['typefalse'], res['typevalfalse'], res['applicationvalfalse'], 'XIVO_SCHEDULE_TYPEFALSE', 'XIVO_SCHEDULE_TYPEVAL1FALSE', 'XIVO_SCHEDULE_TYPEVAL2FALSE')

agid.register(schedule)

File: web-interface/tpl/www/bloc/service/ipbx/asterisk/dialaction/schedule.php
<?php

#
# XiVO Web-Interface
# Copyright (C) 2006-2009  Proformatique <technique@proformatique.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

$form = &$this->get_module('form');
$url = &$this->get_module('url');
$dhtml = &$this->get_module('dhtml');

$element = $this->get_var('element');
$list = $this->get_var('destination_list','schedule');
$event = $this->get_var('event');

$linked = $this->get_var('dialaction',$event,'linked');
$action = $this->get_var('dialaction',$event,'action');

if(empty($list) === false):
	echo	'<div id="fd-dialaction-'.$event.'-schedule-actiontype" class="b-nodisplay">',
		$form->select(array('desc'	=> $this->bbf('fm_dialaction_schedule-actionarg1'),
				    'name'	=> 'dialaction['.$event.'][actionarg1]',
				    'labelid'	=> 'dialaction-'.$event.'-schedule-actionarg1',
				    'key'	=> 'identity',
				    'altkey'	=> 'id',
				    'invalid'	=> ($linked === false && $action === 'schedule'),
				    'default'	=> $element['dialaction']['actionarg1']['default'],
				    'selected'	=> $this->get_var('dialaction',$event,'schedule','actionarg1')),
			      $list);

	if($event === 'voicemenuflow'):
		echo	$form->button(array('name'	=> 'add-defapplication-schedule',
					    'id'	=> 'it-add-defapplication-schedule',
					    'value'	=> $this->bbf('fm_bt-add')),
				      'onclick="xivo_ast_defapplication_schedule(\''.$dhtml->escape($event).'\',\'it-voicemenu-flow\');"');
	elseif($event === 'voicemenuevent'):
		echo	$form->button(array('name'	=> 'select-defapplication-schedule',
					    'id'	=> 'it-select-defapplication-schedule',
					    'value'	=> $this->bbf('fm_bt-select')),
				      'onclick="xivo_ast_voicemenuevent_defapplication(\'schedule\');"');
	endif;
	echo	'</div>';
else:
	echo	'<div id="fd-dialaction-'.$event.'-schedule-actiontype" class="txt-center b-nodisplay">';
	if($this->get_var('dialaction_from') === 'schedule'):
		echo	$this->bbf('dialaction_no-schedule');
	else:
		echo	$url->href_html($this->bbf('create_schedule'),'service/ipbx/call_management/schedule','act=add');
	endif;
	echo	'</div>';
endif;

?>
